File: fast_sync_typescript_service_sdk/packages/fast-sync-core/src/core/implementation/service/sync_manager.ts
import { SyncOperationMetadata } from "../../abstraction/models/Sync_operation_metadata";
import { SyncPayload } from "../../abstraction/models/Sync_payload";
import { SyncOperationEnum } from "../../abstraction/metadata/ISync_operation";
import { ISyncManager } from "../../abstraction/service/ISync_manager";
import { FastSync } from "../fast_sync";
import { isEmptyString } from "../utils/utils";
import { SyncZoneRestrictionEnum } from "../../abstraction/models/Sync_zone_restriction";

export class SyncManager implements ISyncManager {
    constructor(){}
     
    async processPush(payload: SyncPayload){
        payload = SyncPayload.create(payload);
        this.validateMetadataSyncZones(payload.operationMetadata)
        let objectTypes = payload.getSynckedTypes()
        for(const type of objectTypes){
            let objects =  payload.getObjectsForType(type);
            let newObjects = objects.filter(obj => obj?.metadata?.syncOperation == SyncOperationEnum.add)
            let updatedObjects = objects.filter(obj => obj?.metadata?.syncOperation == SyncOperationEnum.update)
            let deletedObjects = objects.filter(obj => obj?.metadata?.syncOperation == SyncOperationEnum.delete)
            let objectRepository = FastSync.getInstance().getObjectRepository(type);
            if(newObjects.length > 0)
                await objectRepository.addMany(newObjects, payload.getTypeMetadata(type));
            if(updatedObjects.length > 0)
                await objectRepository.updateMany(updatedObjects, payload.getTypeMetadata(type));
            if(deletedObjects.length > 0)
                await objectRepository.removeMany(deletedObjects, payload.getTypeMetadata(type));
        }
    }
    
    async processPull(metadata: SyncOperationMetadata): Promise<SyncPayload> {
        metadata = SyncOperationMetadata.create(metadata)
        this.validateMetadataSyncZones(metadata)
        let syncPayload : SyncPayload = new SyncPayload();
        let requestedTypes = metadata.getSynckedTypes();
        let fastSync = FastSync.getInstance();
        for(const type of requestedTypes){
            let objectRepository = fastSync.getObjectRepository(type); 
            let typeMetadata = metadata.getTypeMetadata(type);
            let objects = await objectRepository.fetchMany(metadata.getTypeMetadata(type));
            await syncPayload.pushObjects(type, objects,  typeMetadata.computeSyncZone(fastSync.getSyncZoneConfiguration(type)));
        }
        return syncPayload;
    }

    processSync(metadata: SyncOperationMetadata): SyncPayload{
        throw new Error()
    }

    private validateMetadataSyncZones(operatationMetadata: SyncOperationMetadata){
        let objectTypes = operatationMetadata.getSynckedTypes();
        for(const type of objectTypes){
            let typeMetadata =  operatationMetadata.getTypeMetadata(type);
            if(isEmptyString(typeMetadata.getSyncZone()) && SyncZoneRestrictionEnum.restricted == FastSync.getInstance().getSyncZoneConfiguration(type)){
                throw new Error("type" + type + "is configured to be restricted, but no syncZoneAttribute was provided in the sync operatation metadata")
            }
        }
   } 
}

File: fast_sync_dotnet_service/fast-sync-core/implementation/service/SyncManager.cs
using fast_sync_core.abstraction.data;
using fast_sync_core.implementation.data;
using Microsoft.Extensions.Logging.Abstractions;
using System.Reflection;


namespace fast_sync_core.implementation
{
    public class SyncManager : ISyncManager
    {
        public SyncManager() { }

        public async Task ProcessPush(SyncPayload payload)
        {
            payload = SyncPayload.Create(payload);
            ValidateMetadataSyncZones(payload.OperationMetadata);
            var objectTypes = payload.GetSyncedTypes();
            foreach (var type in objectTypes)
            {
                List<object> objects = payload.GetObjectsForType(type);
                object objectRepository = FastSync.GetObjectRepository<SyncableObject>(type);
                JsonSyncPayloadObjects jsonSyncPayloadObjects = new JsonSyncPayloadObjects().BuildJsonSyncPayload(objects);

                if (jsonSyncPayloadObjects.AddedNew.Any())
                {
                    var result = await _repositoryExecutor(repository: objectRepository, methodName: "AddMany", inputs:[ jsonSyncPayloadObjects.AddedNew, payload.GetTypeMetadata(type) ]);
                }
                if (jsonSyncPayloadObjects.Updated.Any())
                {
                    var result = await _repositoryExecutor(repository: objectRepository, methodName: "UpdateMany", inputs: [jsonSyncPayloadObjects.Updated, payload.GetTypeMetadata(type)]);
                }
                if (jsonSyncPayloadObjects.Deleted.Any())
                {
                    var result = await _repositoryExecutor(repository: objectRepository, methodName: "RemoveMany", inputs: [jsonSyncPayloadObjects.Deleted, payload.GetTypeMetadata(type)]);
                }
            }
        }

        public async Task<SyncPayload> ProcessPull(SyncOperationMetadata metadata)
        {
            metadata = SyncOperationMetadata.Create(metadata);
            ValidateMetadataSyncZones(metadata);
            var syncPayload = new SyncPayload();
            var requestedTypes = metadata.GetSyncedTypes();
            var fastSync = FastSync.GetInstance();
            foreach (var type in requestedTypes)
            {
                object objectRepository =  FastSync.GetObjectRepository<SyncableObject>(type);
                var typeMetadata = metadata.GetTypeMetadata(type);
                object? result = await _repositoryExecutor(repository: objectRepository, methodName: "FetchMany", inputs: [metadata.GetTypeMetadata(type)]);
                List<SyncableObject> collection = result == null ? new List<SyncableObject>() : new List<SyncableObject>((IEnumerable<SyncableObject>)result);
                syncPayload.PushObjects(type, collection, typeMetadata.ComputeSyncZone(fastSync.GetSyncZoneConfiguration(type)));
            }
            return syncPayload;
        }

        private async Task<object?> _repositoryExecutor(object repository, string methodName, object[] inputs)
        {
            object? result = null;
            MethodInfo? addManyInfo = repository.GetType().GetMethod(methodName);
            if (addManyInfo != null)
            {
                Task? addManyTask = addManyInfo.Invoke(repository, inputs) as Task;
                if (addManyTask != null)
                {
                    await addManyTask;
                    var resultProperty = addManyTask.GetType().GetProperty("Result");
                    if (resultProperty != null)
                    {
                        result = resultProperty.GetValue(addManyTask);
                    }
                }
            }
            return result;
        }


        private void ValidateMetadataSyncZones(SyncOperationMetadata operationMetadata)
        {
            var objectTypes = operationMetadata.GetSyncedTypes();
            foreach (var type in objectTypes)
            {
                var typeMetadata = operationMetadata.GetTypeMetadata(type);
                if (String.IsNullOrEmpty(typeMetadata.GetSyncZone()) && FastSync.GetInstance().GetSyncZoneConfiguration(type) == SyncZoneRestrictionEnum.Restricted)
                {
                    throw new InvalidOperationException($"Type {type} is configured to be restricted, but no syncZoneAttribute was provided in the sync operation metadata");
                }
            }
        }
    }
}
